From 26c5ecf8b547575ede7190d44105eec50e696821 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:52:37 -0400
Subject: [tools] lib/traceevent: Unregister handler when function plugin is unloaded

Message-id: <1408953210-26343-270-git-send-email-jolsa@redhat.com>
Patchwork-id: 89015
O-Subject: [PATCH RHEL7.1 BZ1133083 269/322] tools lib traceevent: Unregister handler when function plugin is unloaded
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit ac668c7b7d87c1895faf196a337b5ddfd32a0e21
Author: Namhyung Kim <namhyung@kernel.org>
Date: Thu Jan 16 11:31:09 2014 +0900

description
===========
The function handler should be unregistered when the plugin is unloaded
otherwise it'll try to access invalid memory.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/lib/traceevent/plugin_function.c b/tools/lib/traceevent/plugin_function.c
index 3946148..80ba4ff 100644
--- a/tools/lib/traceevent/plugin_function.c
+++ b/tools/lib/traceevent/plugin_function.c
@@ -148,6 +148,9 @@ void PEVENT_PLUGIN_UNLOADER(struct pevent *pevent)
 {
  int i, x;
 
+ pevent_unregister_event_handler(pevent, -1, "ftrace", "function",
+     function_handler, NULL);
+
  for (i = 0; i <= cpus; i++) {
   for (x = 0; x < fstack[i].size && fstack[i].stack[x]; x++)
    free(fstack[i].stack[x]);
-- 
1.7.1